zero/rect_grid.c

```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
   }
 }
 
+GKYL_CU_DH
+void gkyl_rect_grid_find_cell(struct gkyl_rect_grid *grid, const double *point, bool pickLower, const int **knownIdx, int *cellIdx){
+  int nDim = grid->ndim;
+  int searchNum = 0;
+  int searchDim[GKYL_MAX_DIM], *idxOut;
+  int *dimTrans[GKYL_MAX_DIM];
+  int plusminus[2] = {-1,1};
+  bool allLessEq = true;
+  int iStart[GKYL_MAX_DIM], iEnd[GKYL_MAX_DIM], iMid[GKYL_MAX_DIM], iNew[GKYL_MAX_DIM];
+  int *cells, lowHighCellIdx[2*GKYL_MAX_DIM];
+  double lowerInDir[nDim], upperInDir[nDim];
+  double low, high;
+  
+  for(int d=0; d<nDim; d++){
+    dimTrans[d] = (int*)malloc(1*sizeof(int));
+    if(knownIdx[d]==NULL){
+      searchDim[searchNum] = d;
+      *dimTrans[d] = searchNum;
+      searchNum = searchNum + 1;
+    }else{
+      dimTrans[d] = NULL;      
+      cellIdx[d] = *knownIdx[d];
+      low = grid->lower[d]+(*knownIdx[d]-1)*grid->dx[d];
+      high = grid->lower[d]+(*knownIdx[d])*grid->dx[d];
+      assert(low<point[d]&&high>point[d]);
+    }
+  }
+
+  cells = grid -> cells;
+  for(int d=0; d<searchNum; d++){
+    iStart[d] = 1;
+    iEnd[d] = cells[searchDim[d]];
+    iMid[d] = 0;
+    iNew[d] = 0;
+  }
+  /* Below we use a binary search. That is, if the i-th coordinate of the point in
+   * question is below(above) the ith-coordinate of the current mid-point, we search
+   * the lower(upper) half along that direction in the next iteration.
+   */
+  while (allLessEq){
+    for(int d=0; d<searchNum; d++){
+      iMid[d] = (iStart[d]+iEnd[d])/2;//Integer division intentional
+    }
+
+    if(isInCell(grid, point, iMid, dimTrans, knownIdx)){
+      //Check if neighboring cells also contain this point.
+      for(int k=0; k<searchNum; k++){
+	iNew[k] = iMid[k];
+	lowHighCellIdx[searchDim[k]] = iMid[k];
+	lowHighCellIdx[nDim+searchDim[k]] = iMid[k];
+      }
+      for(int i=0; i<searchNum; i++){	
+	for(int j=0; j<2; j++){
+	  iNew[i] = fmax(fmin(iMid[i] + plusminus[j],cells[i]),0);
+	  if(isInCell(grid, point, iNew, dimTrans, knownIdx)){
+	    lowHighCellIdx[j*nDim+searchDim[i]] = iNew[i];	    
+	  }
+	}
+	iNew[i] = iMid[i];
+      }
+      if(pickLower){
+	for(int d=0; d<searchNum; d++){
+	  cellIdx[searchDim[d]] = lowHighCellIdx[searchDim[d]];
+	}
+      }else{
+	for(int d=0; d<searchNum; d++){
+	  cellIdx[searchDim[d]] = lowHighCellIdx[nDim+searchDim[d]];
+	}
+      }
+      break;
+    }else{
+      InDir(grid, iMid, dimTrans, knownIdx, lowerInDir, upperInDir);
+      for(int d=0; d<searchNum; d++){
+	if(point[searchDim[d]] < lowerInDir[searchDim[d]]){
+	  iEnd[d] = iMid[d]-1;
+	}else if(point[searchDim[d]] > upperInDir[searchDim[d]]){
+	  iStart[d] = iMid[d]+1;
+	}
+      }
+    }
+
+    for(int d=0; d<searchNum; d++){
+      if(iStart[d]>iEnd[d]){
+	allLessEq = false;
+	break;
+      }
+    }
+  }
+
+}
+
+GKYL_CU_DH
+bool isInCell(const struct gkyl_rect_grid *grid, const double *pIn, int *iIn, int *dimTrans[1], const int **knownIdx){
```

Comment:
please add some (very brief) comments describing this function and its inputs/outputs.